src/mmap_storage.cpp

```diff
@@ -160,8 +160,9 @@ error_code translate_error(std::error_code const& err, bool const write)
 			if (fs.pad_file_at(i)) continue;
 
 			download_priority_t const old_prio = m_file_priority[i];
-			download_priority_t new_prio = prio[i];
-			if (old_prio == dont_download && new_prio != dont_download)
+			m_file_priority[i] = prio[i];
```

Comment:
the issue of updating the priority this early is that if there's a failure, it will be left incorrect. This was (probably) the main reason it was done at the end earlier. Perhaps it doesn't matter so much. If it fails, ideally the user would be able to retry, but if we change the priority to non-zero, this logic won't re-trigger.

Reply:
Also, I think it's really important to document the reason for updating the priority earlier (just like it would have been good to document that updating it even in failure cases is problematic

Reply:
> the issue of updating the priority this early is that if there's a failure, it will be left incorrect.

I thought about that. But it seems unavoidable anyway, if we want `open_file` to work right. (Beside changing `open_file_impl` not to use `m_file_priority` somehow.)

Arguably we could also move all `prio = m_file_priority;` to the top. As I understand it, a caller could read the mutated state back.

Reply:
> Arguably we could also move all `prio = m_file_priority;` to the top.

Ok, I see you already came up with the same idea. 😅

Reply:
`prio` is owned by the disk job at this point. however, in the response, we pass back the *actual* priorities that ended up being set, and we use the same vector for that